File: haskell/day15_part2_2015.hs

import Control.Monad (forM_)
import Data.List (intercalate)
import Data.Maybe (mapMaybe)
import System.IO (readFile)

data Ingredient = Ingredient { name :: String, capacity :: Int, durability :: Int, flavor :: Int, texture :: Int, calories :: Int }

main :: IO ()
main = do
    contents <- readFile "input.txt"
    let ingredients = parseIngredients contents
    print $ findMaxScore ingredients 100 500

parseIngredients :: String -> [Ingredient]
parseIngredients = mapMaybe parseLine . lines
  where
    parseLine line = case words line of
        [n, _, c, _, d, _, f, _, t, _, cal] -> Just $ Ingredient n (read $ init c) (read $ init d) (read $ init f) (read $ init t) (read cal)
        _ -> Nothing

findMaxScore :: [Ingredient] -> Int -> Int -> Int
findMaxScore ingredients totalTeaspoons targetCalories = maximum $ calculateScores ingredients totalTeaspoons targetCalories

calculateScores :: [Ingredient] -> Int -> Int -> [Int]
calculateScores ingredients totalTeaspoons targetCalories = 
    [ score | tsp <- combinations (length ingredients) totalTeaspoons, 
              calculateCalories ingredients tsp == targetCalories,
              let score = calculateScore ingredients tsp ]

combinations :: Int -> Int -> [[Int]]
combinations n total = go n total
  where
    go 1 total = [[total]]
    go k total = [x : xs | x <- [0..total], xs <- go (k-1) (total - x)]

calculateScore :: [Ingredient] -> [Int] -> Int
calculateScore ingredients teaspoons = 
    let (cap, dur, fla, tex) = foldl (\(c, d, f, t) (i, tsp) -> (c + capacity (ingredients !! i) * tsp, d + durability (ingredients !! i) * tsp, f + flavor (ingredients !! i) * tsp, t + texture (ingredients !! i) * tsp)) (0, 0, 0, 0) (zip [0..] teaspoons)
    in max 0 (cap * dur * fla * tex)

calculateCalories :: [Ingredient] -> [Int] -> Int
calculateCalories ingredients teaspoons = sum $ zipWith (\i tsp -> calories (ingredients !! i) * tsp) [0..] teaspoons

File: fsharp/day15_part2_2015.fs

open System
open System.IO

type Ingredient = {
    capacity   : int
    durability : int
    flavor     : int
    texture    : int
    calories   : int
}

let parseLine (line:string) =
    let parts = line.Split([|' '; ':'; ','|], StringSplitOptions.RemoveEmptyEntries)
    {
        capacity   = int parts.[2]
        durability = int parts.[4]
        flavor     = int parts.[6]
        texture    = int parts.[8]
        calories   = int parts.[10]
    }

let ingredients =
    File.ReadAllLines("input.txt")
    |> Array.map parseLine

let totalTeaspoons = 100
let targetCalories = 500
let n = ingredients.Length
let teaspoons = Array.zeroCreate<int> n

let score (caps:int) (durs:int) (flas:int) (texs:int) =
    let caps = if caps < 0 then 0 else caps
    let durs = if durs < 0 then 0 else durs
    let flas = if flas < 0 then 0 else flas
    let texs = if texs < 0 then 0 else texs
    caps * durs * flas * texs

let rec dfs idx remaining =
    if idx = n - 1 then
        teaspoons.[idx] <- remaining
        let cal =
            Array.fold2 (fun acc ing tsp -> acc + ing.calories * tsp) 0 ingredients teaspoons
        if cal = targetCalories then
            let cap = Array.fold2 (fun acc ing tsp -> acc + ing.capacity * tsp) 0 ingredients teaspoons
            let dur = Array.fold2 (fun acc ing tsp -> acc + ing.durability * tsp) 0 ingredients teaspoons
            let fla = Array.fold2 (fun acc ing tsp -> acc + ing.flavor * tsp) 0 ingredients teaspoons
            let tex = Array.fold2 (fun acc ing tsp -> acc + ing.texture * tsp) 0 ingredients teaspoons
            score cap dur fla tex
        else 0
    else
        let mutable best = 0
        for i in 0 .. remaining do
            teaspoons.[idx] <- i
            let cur = dfs (idx + 1) (remaining - i)
            if cur > best then best <- cur
        best

[<EntryPoint>]
let main _ =
    let result = dfs 0 totalTeaspoons
    printfn "%d" result
    0
